grace/models/optimiser.py

```diff
@@ -141,15 +141,19 @@ def optimise_graph(
     # build a set of false positive hypotheses
     for i, n_dict in graph.nodes.data():
         hypotheses.append(
-            Hypothesis(i=i, j=None, rho=n_dict[GraphAttrs.NODE_PREDICTION])
+            Hypothesis(
+                i=i, j=None, rho=n_dict[GraphAttrs.NODE_PREDICTION][1][0]
+            )
         )
         hypotheses.append(
-            Hypothesis(i=None, j=i, rho=n_dict[GraphAttrs.NODE_PREDICTION])
+            Hypothesis(
+                i=None, j=i, rho=n_dict[GraphAttrs.NODE_PREDICTION][1][0]
+            )
         )
     # build a set of link hypotheses
     for i, j, e_dict in graph.edges.data():
         hypotheses.append(
-            Hypothesis(i=i, j=j, rho=e_dict[GraphAttrs.EDGE_PREDICTION])
+            Hypothesis(i=i, j=j, rho=e_dict[GraphAttrs.EDGE_PREDICTION][1][1])
```

Comment:
Again, do you need the array here? It seems like a simple probability would do

Reply:
Nodes and edges need probability of a different class for the optimiser to work, as stated in the documentation.

Reply:
Sure, but you're indexing by that anyway with the `GraphAttrs` key. I think you've clarified that what you want to store is the probability distribution over the labels - which seems to make sense. I would just make that clear in the documentation and change how you store it in the edge attributes, as noted above

Reply:
All implemented in the latest update - thanks for the feedback & helpful suggestions on how to improve the code!